dtm0/clk_src.c

```diff
@@ -0,0 +1,165 @@
+	int  (*cso_cmp)(const struct m0_dtm0_ts *left,
+			const struct m0_dtm0_ts *right);
+	void (*cso_now)(struct m0_dtm0_clk_src *cs, struct m0_dtm0_ts *ts);
+	int  (*cso_observed)(struct m0_dtm0_clk_src  *cs,
+			     const struct m0_dtm0_ts *other_ts);
+};
+
+/** See M0_DTM0_CS_PHYS for details */
+static const struct m0_dtm0_clk_src_ops cs_phys_ops;
+
+M0_INTERNAL void m0_dtm0_clk_src_init(struct m0_dtm0_clk_src *cs,
+				      enum m0_dtm0_cs_types   type)
+{
+	M0_ENTRY();
+	M0_PRE(M0_IN(type, (M0_DTM0_CS_PHYS)));
+	m0_mutex_init(&cs->cs_phys_lock);
+	cs->cs_last = M0_DTM0_TS_MIN;
+	cs->cs_ops = &cs_phys_ops;
+	M0_LEAVE();
+}
+
+M0_INTERNAL void m0_dtm0_clk_src_fini(struct m0_dtm0_clk_src *cs)
+{
+	M0_ENTRY();
+	M0_PRE(cs);
+	M0_PRE(M0_IN(cs->cs_ops, (&cs_phys_ops)));
+	m0_mutex_fini(&cs->cs_phys_lock);
+	cs->cs_last = M0_DTM0_TS_INIT;
+	cs->cs_ops = NULL;
+	M0_LEAVE();
+}
+
+M0_INTERNAL
+enum m0_dtm0_ts_ord m0_dtm0_ts_cmp(const struct m0_dtm0_clk_src *cs,
+				   const struct m0_dtm0_ts      *left,
+				   const struct m0_dtm0_ts      *right)
+{
+	M0_PRE(cs);
+	M0_PRE(m0_dtm0_ts__invariant(left));
+	M0_PRE(m0_dtm0_ts__invariant(right));
+	return cs->cs_ops->cso_cmp(left, right);
+}
+
+M0_INTERNAL void m0_dtm0_clk_src_now(struct m0_dtm0_clk_src *cs,
+				     struct m0_dtm0_ts      *now)
+{
+	M0_PRE(cs);
+	cs->cs_ops->cso_now(cs, now);
+	M0_POST(m0_dtm0_ts__invariant(now));
+}
+
+M0_INTERNAL int m0_dtm0_clk_src_observed(struct m0_dtm0_clk_src  *cs,
+					 const struct m0_dtm0_ts *other_ts)
+{
+	M0_PRE(m0_dtm0_ts__invariant(other_ts));
+	return cs->cs_ops->cso_observed(cs, other_ts);
+}
+
+M0_INTERNAL bool m0_dtm0_ts__invariant(const struct m0_dtm0_ts *ts)
+{
+	const struct m0_dtm0_ts min_ts = M0_DTM0_TS_MIN;
+	const struct m0_dtm0_ts max_ts = M0_DTM0_TS_MAX;
+
+	return _0C(ts->dts_phys >= min_ts.dts_phys) &&
+		_0C(ts->dts_phys <= max_ts.dts_phys);
+}
+
+/* Phys clock implementation */
+
+static enum m0_dtm0_ts_ord cs_phys_cmp(const struct m0_dtm0_ts *left,
+				       const struct m0_dtm0_ts *right)
+{
+	return M0_3WAY(left->dts_phys, right->dts_phys);
+}
+
+static void cs_phys_now(struct m0_dtm0_clk_src *cs, struct m0_dtm0_ts *now)
+{
+	M0_PRE(cs);
+	M0_PRE(M0_IN(cs->cs_ops, (&cs_phys_ops)));
+
+	m0_mutex_lock(&cs->cs_phys_lock);
+
+	now->dts_phys = m0_time_now();
+	if (now->dts_phys > cs->cs_last.dts_phys) {
+		cs->cs_last = *now;
+	} else {
+		cs->cs_last.dts_phys++;
+		*now = cs->cs_last;
+	}
+
+	m0_mutex_unlock(&cs->cs_phys_lock);
+}
+
+static int cs_phys_observed(struct m0_dtm0_clk_src  *cs,
+			    const struct m0_dtm0_ts *other_ts)
+{
+	struct m0_dtm0_ts now = M0_DTM0_TS_INIT;
+	int               rc;
+
+	M0_PRE(m0_dtm0_ts__invariant(other_ts));
+
+	m0_dtm0_clk_src_now(cs, &now);
+
+	rc = m0_dtm0_ts_cmp(cs, &now, other_ts);
+
+	/*
+	 * Assumption:
+	 * It is possible to observe an even that happened in the past.
+	 * The other cases (present, future) are not allowed, and they
+	 * indicate a significant clock drift.
```

Comment:
What if the drift is on par with the maximal allowed duration of a transient failure? In this case, we might have troubles with log pruning or with REDOs "beating" new (from real-world perspective) transactions. That's probably the only case where we can consider the drift to be significant. However, I believe NTP will help us to avoid it. And even if it does not help us, detection of such cases should be up to the user, the clock source should not make such decisions. And moreover we did not use CS.OBSERVED outside of the UT code. That's why I just removed CS.OBSERVED and the related code in https://github.com/Seagate/cortx-motr/pull/561/commits/0ec44d05c4f76f18b53b4e4b77a0073561ae2097.